landlab/components/concentration_tracker/concentration_tracker_for_diffusion.py

```diff
@@ -0,0 +1,521 @@
+        else:
+            if concentration_in_bedrock < 0:
+                raise ValueError("Concentration in bedrock cannot be negative.")
+
+    @property
+    def C_init(self):
+        """Initial concentration in soil/sediment (kg/m^3)."""
+        return self._C_init
+
+    @property
+    def C_br(self):
+        """Concentration in bedrock (kg/m^3)."""
+        return self._C_br
+
+    @property
+    def P(self):
+        """Rate of local production (kg/m^3/yr)."""
+        return self._P
+
+    @property
+    def D(self):
+        """Rate of local decay (kg/m^3/yr)."""
+        return self._D
+
+    @C_init.setter
+    def C_init(self, new_val):
+        self._C_init = return_array_at_node(self._grid, new_val)
+
+    @C_br.setter
+    def C_br(self, new_val):
+        self._C_br = return_array_at_node(self._grid, new_val)
+
+    @P.setter
+    def P(self, new_val):
+        self._P = return_array_at_node(self._grid, new_val)
+
+    @D.setter
+    def D(self, new_val):
+        self._D = return_array_at_node(self._grid, new_val)
+
+    def concentration(self, dt):
+        """Calculate change in concentration for a time period 'dt'.
+
+        Parameters
+        ----------
+
+        dt: float (time)
+            The imposed timestep.
+        """
+
+        # Define concentration at previous timestep
+        C_old = self._concentration.copy()
+
+        # Map concentration from nodes to links (following soil flux direction)
+        # Does this overwrite fixed-value/gradient links?
+        self._C_links = map_value_at_max_node_to_link(
+            self._grid, "topographic__elevation", "sediment_property__concentration"
+        )
+        # Replace values with zero for all INACTIVE links
+        self._C_links[self._grid.status_at_link == LinkStatus.INACTIVE] = 0.0
+
+        # Calculate QC at links (sediment flux times concentration)
+        self._grid.at_link["QC"] = (
+            self._grid.at_link["soil__flux"][:] * self._C_links[:]
+        )
+
+        # Calculate flux concentration divergence
+        dQCdx = self._grid.calc_flux_div_at_node(self._grid.at_link["QC"])
+
+        # Calculate other components of mass balance equation
+        with np.errstate(divide="ignore", invalid="ignore"):
+            C_local = C_old * (self._soil__depth_old / self._soil__depth)
+            C_from_weathering = (
+                self._C_br * (self._soil_prod_rate * dt) / self._soil__depth
+            )
+            Production = (dt * self._P / 2) * (
+                self._soil__depth_old / self._soil__depth + 1
+            )
+            Decay = (dt * self._D / 2) * (self._soil__depth_old / self._soil__depth + 1)
+
+            # Calculate concentration
+            self._concentration[:] = (
+                C_local
+                + C_from_weathering
+                + (dt / self._soil__depth) * (-dQCdx)
+                + Production
+                - Decay
+            )
```

Comment:
Instead of of changing the `np.errstate` could you, instead, be more explicit about what to do with nodes with zero soil depth? Maybe something along the lines of,
```python
is_soil = self._soil__depth > 0.0

C_local = np.divide(
    C_old * self._soil__depth_old, self._soil__depth, where=is_soil
)
C_from_weathering = np.divide(
    self._C_br * (self._soil_prod_rate * dt), self._soil__depth, where=is_soil
)
Production = (dt * self._P / 2) * (
    np.divide(self._soil__depth_old, self._soil__depth, where=is_soil) + 1
)
Decay = (dt * self._D / 2) * (
    np.divide(self._soil__depth_old, self._soil__depth, where=is_soil) + 1
)

self._concentration[:] = (
    C_local
    + C_from_weathering
    + (dt / self._soil__depth) * (-dQCdx)
    + Production
    - Decay
)

self._concentration[~is_soil] = 0.0
```